mj-carousel

LiamRiddell opened the pull request.

LiamRiddell commented on Mjml.Net/Components/Body/CarouselComponent.cs:
```diff
@@ -0,0 +1,453 @@
+            }
+        }
+
+        private void HeadStyle(IHtmlRenderer renderer, GlobalContext context)
```
@SebastianStehle This is horrible. Any ideas on how we could better approach this?

LiamRiddell commented on Mjml.Net/Components/Body/CarouselComponent.cs:
```diff
@@ -0,0 +1,453 @@
+
+        public override void Render(IHtmlRenderer renderer, GlobalContext context)
+        {
+            CarouselID = Guid.NewGuid().ToString();
```
@SebastianStehle, I'm using GUID but on the original it uses 6 random bytes. Should we change?

SebastianStehle replied:
No, it is fine, I have removed some attributes like "id" and "for" from the HTML diff.

LiamRiddell commented on Mjml.Net/Components/Body/CarouselImageComponent.cs:
```diff
@@ -0,0 +1,115 @@
+                .Attr("src", Src)
+                .Attr("alt", Alt)
+                .Attr("border", "0")
+                // .Attr("width", containerWidth) NOT SURE
```
@SebastianStehle What's the best way to populate down the container width? I'm a little confused here. I've marked all similar `containerWidth` attributes/styles the same.

SebastianStehle replied:
If a method does not override Measure, the ActualWidth is the same as containerwidth. If you override Measure, you can get the containerWidth from the parentWidth there.

LiamRiddell reviewed (commented):
@SebastianStehle, This component is a horrible component in MJML. There are a few parts that I need some help on. Please see the review comments.

LiamRiddell commented:
@SebastianStehle, I'm going to review all this now. Apologies for delays.

LiamRiddell commented:
@SebastianStehle, We have an issue on this component where by the IDs are generating all the test failures. The problem is that most of them are triggered by classes or names.

```
1 The values of the attributes at div(0) > input(0)[class] are different.
   *   Actual: 'mj-carousel-radio mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-radio mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-radio-1'.
   * Expected: 'mj-carousel-radio mj-carousel-41d58ca8b0b9-radio mj-carousel-41d58ca8b0b9-radio-1'.
 2 The values of the attributes at div(0) > input(0)[name] are different.
   *   Actual: 'mj-carousel-radio-14f99efd-5e7e-4876-b449-ee705e827ff6'.
   * Expected: 'mj-carousel-radio-41d58ca8b0b9'.
 3 The values of the attributes at div(0) > input(1)[class] are different.
   *   Actual: 'mj-carousel-radio mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-radio mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-radio-2'.
   * Expected: 'mj-carousel-radio mj-carousel-41d58ca8b0b9-radio mj-carousel-41d58ca8b0b9-radio-2'.
 4 The values of the attributes at div(0) > input(1)[name] are different.
   *   Actual: 'mj-carousel-radio-14f99efd-5e7e-4876-b449-ee705e827ff6'.
   * Expected: 'mj-carousel-radio-41d58ca8b0b9'.
 5 The values of the attributes at div(0) > input(2)[class] are different.
   *   Actual: 'mj-carousel-radio mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-radio mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-radio-3'.
   * Expected: 'mj-carousel-radio mj-carousel-41d58ca8b0b9-radio mj-carousel-41d58ca8b0b9-radio-3'.
 6 The values of the attributes at div(0) > input(2)[name] are different.
   *   Actual: 'mj-carousel-radio-14f99efd-5e7e-4876-b449-ee705e827ff6'.
   * Expected: 'mj-carousel-radio-41d58ca8b0b9'.
 7 The values of the attributes at div(0) > div(3)[class] are different.
   *   Actual: 'mj-carousel-content mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-content'.
   * Expected: 'mj-carousel-content mj-carousel-41d58ca8b0b9-content'.
 8 The values of the attributes at div(0) > div(3) > a(0)[class] are different.
   *   Actual: 'mj-carousel-thumbnail mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-thumbnail mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-thumbnail-1'.
   * Expected: 'mj-carousel-thumbnail mj-carousel-41d58ca8b0b9-thumbnail mj-carousel-41d58ca8b0b9-thumbnail-1 '.
 9 The values of the attributes at div(0) > div(3) > a(1)[class] are different.
   *   Actual: 'mj-carousel-thumbnail mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-thumbnail mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-thumbnail-2'.
   * Expected: 'mj-carousel-thumbnail mj-carousel-41d58ca8b0b9-thumbnail mj-carousel-41d58ca8b0b9-thumbnail-2 '.
 10 The values of the attributes at div(0) > div(3) > a(2)[class] are different.
   *   Actual: 'mj-carousel-thumbnail mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-thumbnail mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-thumbnail-3'.
   * Expected: 'mj-carousel-thumbnail mj-carousel-41d58ca8b0b9-thumbnail mj-carousel-41d58ca8b0b9-thumbnail-3 '.
 11 The values of the attributes at div(0) > div(3) > table(3) > tbody(0) > tr(0) > td(0)[class] are different.
   *   Actual: 'mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-icons-cell'.
   * Expected: 'mj-carousel-41d58ca8b0b9-icons-cell'.
 12 The values of the attributes at div(0) > div(3) > table(3) > tbody(0) > tr(0) > td(2)[class] are different.
   *   Actual: 'mj-carousel-14f99efd-5e7e-4876-b449-ee705e827ff6-icons-cell'.
   * Expected: 'mj-carousel-41d58ca8b0b9-icons-cell'.
```

LiamRiddell commented:
@SebastianStehle, additionally, what's the best approach for testing the HeadStyle?

SebastianStehle commented:
About IDs:
See the comments in this PR: https://github.com/SebastianStehle/mjml-net/pull/94

About Styles:
You can render styles only with 

```
<mjml-test body="false">
   ... Your component
</mjml>
```

LiamRiddell commented:
> About IDs: See the comments in this PR: #94
> 
> About Styles: You can render styles only with
> 
> ```
> <mjml-test body="false">
>    ... Your component
> </mjml>
> ```

ID's are working now. Thank you :) It looks like we may have a bug with the mj-test component. Here the example:

```
<mjml-test head=""true"" body=""false"">
    <mj-carousel>
        <mj-carousel-image src=""https://www.mailjet.com/wp-content/uploads/2016/11/ecommerce-guide.jpg"" />
        <mj-carousel-image src=""https://www.mailjet.com/wp-content/uploads/2016/09/3@1x.png"" />
        <mj-carousel-image src=""https://www.mailjet.com/wp-content/uploads/2016/09/1@1x.png"" />
        <mj-carousel-image src=""https://www.mailjet.com/wp-content/uploads/2016/09/1@1x.png"" />
    </mj-carousel>
</mjml-test>
```
This outputs:
```
<!--[if !mso><!-->
<div class="mj-carousel">
  <input type="radio" name="mj-carousel-radio-9dd9de129727" id="mj-carousel-9dd9de129727-radio-1" checked="checked" class="mj-carousel-radio mj-carousel-9dd9de129727-radio mj-carousel-9dd9de129727-radio-1"  style="display:none;mso-hide:all;"/>
  <input type="radio" name="mj-carousel-radio-9dd9de129727" id="mj-carousel-9dd9de129727-radio-2" class="mj-carousel-radio mj-carousel-9dd9de129727-radio mj-carousel-9dd9de129727-radio-2"  style="display:none;mso-hide:all;"/>
  <input type="radio" name="mj-carousel-radio-9dd9de129727" id="mj-carousel-9dd9de129727-radio-3" class="mj-carousel-radio mj-carousel-9dd9de129727-radio mj-carousel-9dd9de129727-radio-3"  style="display:none;mso-hide:all;"/>
  <input type="radio" name="mj-carousel-radio-9dd9de129727" id="mj-carousel-9dd9de129727-radio-4" class="mj-carousel-radio mj-carousel-9dd9de129727-radio mj-carousel-9dd9de129727-radio-4"  style="display:none;mso-hide:all;"/>
  <div class="mj-carousel-content mj-carousel-9dd9de129727-content"  style="display:table;width:100%;table-layout:fixed;text-align:center;font-size:0px;">
    <a href="#1" target="_blank" class="mj-carousel-thumbnail mj-carousel-9dd9de129727-thumbnail mj-carousel-9dd9de129727-thumbnail-1"  style="border:2px solid transparent;border-radius:6px;display:inline-block;overflow:hidden;width:110px;">
      <label for="mj-carousel-9dd9de129727-radio-1">
        <img src="https://www.mailjet.com/wp-content/uploads/2016/11/ecommerce-guide.jpg" width="110" style="display:block;width:100%;height:auto;"/>
      </label>
    </a>
    <a href="#2" target="_blank" class="mj-carousel-thumbnail mj-carousel-9dd9de129727-thumbnail mj-carousel-9dd9de129727-thumbnail-2"  style="border:2px solid transparent;border-radius:6px;display:inline-block;overflow:hidden;width:110px;">
      <label for="mj-carousel-9dd9de129727-radio-2">
        <img src="https://www.mailjet.com/wp-content/uploads/2016/09/3@1x.png" width="110" style="display:block;width:100%;height:auto;"/>
      </label>
    </a>
    <a href="#3" target="_blank" class="mj-carousel-thumbnail mj-carousel-9dd9de129727-thumbnail mj-carousel-9dd9de129727-thumbnail-3"  style="border:2px solid transparent;border-radius:6px;display:inline-block;overflow:hidden;width:110px;">
      <label for="mj-carousel-9dd9de129727-radio-3">
        <img src="https://www.mailjet.com/wp-content/uploads/2016/09/1@1x.png" width="110" style="display:block;width:100%;height:auto;"/>
      </label>
    </a>
    <a href="#4" target="_blank" class="mj-carousel-thumbnail mj-carousel-9dd9de129727-thumbnail mj-carousel-9dd9de129727-thumbnail-4"  style="border:2px solid transparent;border-radius:6px;display:inline-block;overflow:hidden;width:110px;">
      <label for="mj-carousel-9dd9de129727-radio-4">
        <img src="https://www.mailjet.com/wp-content/uploads/2016/09/1@1x.png" width="110" style="display:block;width:100%;height:auto;"/>
      </label>
    </a>
    <table border="0" cell-padding="0" cell-spacing="0" width="100%" role="presentation" class="mj-carousel-main"  style="caption-side:top;display:table-caption;table-layout:fixed;width:100%;">
      <tbody>
        <tr>
          <td class="mj-carousel-9dd9de129727-icons-cell"  style="font-size:0px;display:none;mso-hide:all;padding:0px;">
            <div class="mj-carousel-previous-icons"  style="display:none;mso-hide:all;">
              <label for="mj-carousel-9dd9de129727-radio-1" class="mj-carousel-previous mj-carousel-previous-1">
                <img src="https://i.imgur.com/xTh3hln.png" alt="previous" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
              <label for="mj-carousel-9dd9de129727-radio-2" class="mj-carousel-previous mj-carousel-previous-2">
                <img src="https://i.imgur.com/xTh3hln.png" alt="previous" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
              <label for="mj-carousel-9dd9de129727-radio-3" class="mj-carousel-previous mj-carousel-previous-3">
                <img src="https://i.imgur.com/xTh3hln.png" alt="previous" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
              <label for="mj-carousel-9dd9de129727-radio-4" class="mj-carousel-previous mj-carousel-previous-4">
                <img src="https://i.imgur.com/xTh3hln.png" alt="previous" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
            </div>
          </td>
          <td style="padding:0px;">
            <div class="mj-carousel-images">
              <div class="mj-carousel-image mj-carousel-image-1">
                <img src="https://www.mailjet.com/wp-content/uploads/2016/11/ecommerce-guide.jpg" border="0" width="600" style="border-radius:6px;display:block;width:600px;max-width:100%;height:auto;"/>
              </div>
              <div class="mj-carousel-image mj-carousel-image-2"  style="display:none;mso-hide:all;">
                <img src="https://www.mailjet.com/wp-content/uploads/2016/09/3@1x.png" border="0" width="600" style="border-radius:6px;display:block;width:600px;max-width:100%;height:auto;"/>
              </div>
              <div class="mj-carousel-image mj-carousel-image-3"  style="display:none;mso-hide:all;">
                <img src="https://www.mailjet.com/wp-content/uploads/2016/09/1@1x.png" border="0" width="600" style="border-radius:6px;display:block;width:600px;max-width:100%;height:auto;"/>
              </div>
              <div class="mj-carousel-image mj-carousel-image-4"  style="display:none;mso-hide:all;">
                <img src="https://www.mailjet.com/wp-content/uploads/2016/09/1@1x.png" border="0" width="600" style="border-radius:6px;display:block;width:600px;max-width:100%;height:auto;"/>
              </div>
            </div>
          </td>
          <td class="mj-carousel-9dd9de129727-icons-cell"  style="font-size:0px;display:none;mso-hide:all;padding:0px;">
            <div class="mj-carousel-next-icons"  style="display:none;mso-hide:all;">
              <label for="mj-carousel-9dd9de129727-radio-1" class="mj-carousel-next mj-carousel-next-1">
                <img src="https://i.imgur.com/os7o9kz.png" alt="next" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
              <label for="mj-carousel-9dd9de129727-radio-2" class="mj-carousel-next mj-carousel-next-2">
                <img src="https://i.imgur.com/os7o9kz.png" alt="next" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
              <label for="mj-carousel-9dd9de129727-radio-3" class="mj-carousel-next mj-carousel-next-3">
                <img src="https://i.imgur.com/os7o9kz.png" alt="next" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
              <label for="mj-carousel-9dd9de129727-radio-4" class="mj-carousel-next mj-carousel-next-4">
                <img src="https://i.imgur.com/os7o9kz.png" alt="next" width="44" style="display:block;height:auto;width:44px;">
                </img>
              </label>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
<!--<![endif]-->
<!--[if mso]>
<div class="mj-carousel-image mj-carousel-image-1">
  <img src="https://www.mailjet.com/wp-content/uploads/2016/11/ecommerce-guide.jpg" border="0" width="600" style="border-radius:6px;display:block;width:600px;max-width:100%;height:auto;"/>
</div>
<![endif]-->
<title>
</title>
<style type="text/css">
  @media only screen and (min-width:480px) {
  }
</style>
<style media="screen and (min-width:480px)">
</style>
<style type="text/css">
  .mj-carousel {
    -webkit-user-select: none;
    -moz-user-select: none;
    user-select: none;
  }
  .mj-carousel-9dd9de129727-icons-cell {
    display: table-cell !important;
   width: 44px !important;
  }
  .mj-carousel-radio,
  .mj-carousel-next,
  .mj-carousel-previous {
    display: none !important;
  }
  .mj-carousel-thumbnail,
  .mj-carousel-next,
  .mj-carousel-previous {
    touch-action: manipulation;
  }
  .mj-carousel-9dd9de129727-radio:checked + .mj-carousel-content .mj-carousel-image, 
  .mj-carousel-9dd9de129727-radio:checked + * + .mj-carousel-content .mj-carousel-image, 
  .mj-carousel-9dd9de129727-radio:checked + * + * + .mj-carousel-content .mj-carousel-image, 
  .mj-carousel-9dd9de129727-radio:checked + * + * + * + .mj-carousel-content .mj-carousel-image {
    display: none !important;
  }
  .mj-carousel-9dd9de129727-radio-1:checked + * + * + * + .mj-carousel-content .mj-carousel-image-1, 
  .mj-carousel-9dd9de129727-radio-2:checked + * + * + .mj-carousel-content .mj-carousel-image-2, 
  .mj-carousel-9dd9de129727-radio-3:checked + * + .mj-carousel-content .mj-carousel-image-3, 
  .mj-carousel-9dd9de129727-radio-4:checked + .mj-carousel-content .mj-carousel-image-4 {
    display: block  !important;
  }
  .mj-carousel-previous-icons,
  .mj-carousel-next-icons,
  .mj-carousel-9dd9de129727-radio-1:checked + * + * + * + .mj-carousel-content .mj-carousel-next-2, 
  .mj-carousel-9dd9de129727-radio-2:checked + * + * + .mj-carousel-content .mj-carousel-next-3, 
  .mj-carousel-9dd9de129727-radio-3:checked + * + .mj-carousel-content .mj-carousel-next-4, 
  .mj-carousel-9dd9de129727-radio-4:checked + .mj-carousel-content .mj-carousel-next-1, 
  .mj-carousel-9dd9de129727-radio-1:checked + * + * + * + .mj-carousel-content .mj-carousel-previous-4, 
  .mj-carousel-9dd9de129727-radio-2:checked + * + * + .mj-carousel-content .mj-carousel-previous-1, 
  .mj-carousel-9dd9de129727-radio-3:checked + * + .mj-carousel-content .mj-carousel-previous-2, 
  .mj-carousel-9dd9de129727-radio-4:checked + .mj-carousel-content .mj-carousel-previous-3 {
    display: block  !important;
  }
  .mj-carousel-9dd9de129727-radio-1:checked + * + * + * + .mj-carousel-content .mj-carousel-9dd9de129727-thumbnail-1, 
  .mj-carousel-9dd9de129727-radio-2:checked + * + * + .mj-carousel-content .mj-carousel-9dd9de129727-thumbnail-2, 
  .mj-carousel-9dd9de129727-radio-3:checked + * + .mj-carousel-content .mj-carousel-9dd9de129727-thumbnail-3, 
  .mj-carousel-9dd9de129727-radio-4:checked + .mj-carousel-content .mj-carousel-9dd9de129727-thumbnail-4 {
    border-color: #ccc  !important;
  }
  .mj-carousel-image img + div,
  .mj-carousel-thumbnail img + div {
    display: none !important;
  }
  .mj-carousel-9dd9de129727-thumbnail:hover + * + * + * + .mj-carousel-main .mj-carousel-image, 
  .mj-carousel-9dd9de129727-thumbnail:hover + * + * + .mj-carousel-main .mj-carousel-image, 
  .mj-carousel-9dd9de129727-thumbnail:hover + * + .mj-carousel-main .mj-carousel-image, 
  .mj-carousel-9dd9de129727-thumbnail:hover + .mj-carousel-main .mj-carousel-image {
    display: none !important;
  }
  .mj-carousel-thumbnail:hover {
    border-color: #fead0d !important;
  }
  .mj-carousel-9dd9de129727-thumbnail-1:hover + * + * + * + .mj-carousel-main .mj-carousel-image-1, 
  .mj-carousel-9dd9de129727-thumbnail-2:hover + * + * + .mj-carousel-main .mj-carousel-image-2, 
  .mj-carousel-9dd9de129727-thumbnail-3:hover + * + .mj-carousel-main .mj-carousel-image-3, 
  .mj-carousel-9dd9de129727-thumbnail-4:hover + .mj-carousel-main .mj-carousel-image-4 {
    display: block !important;
  }
  .mj-carousel noinput { display:block !important; }
  .mj-carousel noinput .mj-carousel-image-1 { display: block !important;  }
  .mj-carousel noinput .mj-carousel-arrows,
  .mj-carousel noinput .mj-carousel-thumbnails { display: none !important; }
  [owa] .mj-carousel-thumbnail { display: none !important; }
  @media screen yahoo {
  .mj-carousel-9dd9de129727-icons-cell,
  .mj-carousel-previous-icons,
  .mj-carousel-next-icons {
    display: none !important;
  }
  .mj-carousel-9dd9de129727-radio-1:checked + * + * + * + .mj-carousel-content .mj-carousel-9dd9de129727-thumbnail-1 {
    border-color: transparent;
  }
</style>
<style type="text/css">
</style>
```

I think it's due to the RenderChildren call potentially. I'm still looking into it.

LiamRiddell commented:
@SebastianStehle, Can you check these tests? They worked locally, I added the "Should_render_carousel_icon_width" and now they've all broken. The error is not clear:

```
  Expected: True
  Actual:   False
```

I've debugged the tests myself but it's like they can't get embedded resource. It's really strange...

LiamRiddell commented:
The `diff` enumeration yields no results resulting in a `True` expression into the `Assert.True(!diffs.Any(), FormatDiffs(diffs)`. However, the Fortmat diffs is still being called. When the Assert is true.

SebastianStehle commented:
A simple "ToList()" has helped.

LiamRiddell commented:
I actually thought, I had gone crazy :D

SebastianStehle commented:
As the writer of the last component, you have the honour to merge it and update nuget version to 1.0

LiamRiddell commented:
Once reviewed, I'll merge this :)
